src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -739,6 +741,76 @@ class StringGenerator : public Generator<std::string> {
     const size_t _alphabetLength;
 };
 
+/** `{^ChooseFromDataset:{...}` */
+class RandomStringFromDataset : public Generator<std::string> {
+
+public:
+    RandomStringFromDataset(const Node& node, GeneratorArgs generatorArgs)
+        : _rng{generatorArgs.rng},
+          _id{generatorArgs.actorId},
+          _path{node["path"].maybe<std::string>().value()},
+          _pathLength{_path.size()}{}
+
+    void loadDataset() {
+      {
+        if (_pathLength <= 0) {
+            BOOST_THROW_EXCEPTION(InvalidValueGeneratorSyntax(
+            "ChooseFromDataset requieres non-empty path"));
+            }
+
+        {
+            if (_all_datasets.count(_path) == 0){
+                _ifs.open(_path, std::ifstream::in);
+                if (_ifs.is_open()){
+                    while(std::getline(_ifs, _current_line))
+                    {
+                        // Ignore emtpy lines
+                        if (_current_line.size()!=0){
+                            _all_datasets[_path].push_back(_current_line);
```

Comment:
You could read this in and assemble everything locally, and then put it into the dataset. That feels cleaner to me, but maybe it doesn't make a difference.

Reply:
Ya that's what I would do as well, but there's definitely a matter of personal preference there.

(If it speaks to you, you might also consider a `DataSetHolder` helper class that encapsulates the logic of "load dataset unless loaded and do so thread-safe-ly" and then have a `static DataSetHolder _holder` in the value-generator. This is bike-shedding, however, so please don't feel compelled to do so if you don't want to. Biggest thing imho is removing as much member-variable state as possible and ensuring the locks/mutexes have correct bookkeeping.)